widgets/treeview/treeview.go

```diff
@@ -0,0 +1,818 @@
+package treeview
+
+import (
+	"errors"
+	"fmt"
+	"image"
+	"io"
+	"log"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/mum4k/termdash/cell"
+	"github.com/mum4k/termdash/keyboard"
+	"github.com/mum4k/termdash/mouse"
+	"github.com/mum4k/termdash/private/canvas"
+	"github.com/mum4k/termdash/private/runewidth"
+	"github.com/mum4k/termdash/terminal/terminalapi"
+	"github.com/mum4k/termdash/widgetapi"
+)
+
+// Number of nodes to scroll per mouse wheel event
+const (
+	ScrollStep = 5
+)
+
+// TreeNode represents a node in the treeview.
+type TreeNode struct {
+	ID            string
+	Label         string
+	Level         int
+	Parent        *TreeNode
+	Children      []*TreeNode
+	Value         interface{} // Can hold any data type
+	ShowSpinner   bool
+	OnClick       func() error
+	ExpandedState bool // Unique expanded state for each node
+	SpinnerIndex  int  // Current index for spinner icons
+	mu            sync.Mutex
+}
+
+// SetShowSpinner safely sets the ShowSpinner flag.
+func (node *TreeNode) SetShowSpinner(value bool) {
```

Comment:
It is recommended that the name of the receiver (here `node`) represents the dentity of the type. Consider `tn` or similar instead.

```go
func (tn *TreeNode) ...
```